File: Project1/Baitoan3/c/self_conjugate_partition_k_parts.py
"""
Self-Conjugate Partition Analysis Module
Implements recursive and dynamic programming approaches for counting 
self-conjugate partitions with exactly k parts
"""

from typing import List, Tuple, Dict
from functools import lru_cache


class SelfConjugatePartitions:
    """Class containing algorithms for self-conjugate partition analysis"""
    
    @staticmethod
    def validate_self_conjugate_property(partition_parts: List[int]) -> bool:
        """
        Check if a partition is self-conjugate (equals its transpose)
        
        Args:
            partition_parts: List of partition parts in non-increasing order
            
        Returns:
            True if partition is self-conjugate, false otherwise
        """
        number_of_parts = len(partition_parts)
        conjugate_form = []
        maximum_part = max(partition_parts) if partition_parts else 0
        
        for row_index in range(1, maximum_part + 1):
            column_count = 0
            for part_value in partition_parts:
                if part_value >= row_index:
                    column_count += 1
                else:
                    break
            conjugate_form.append(column_count)
        
        if len(conjugate_form) != number_of_parts:
            return False
        
        for index in range(number_of_parts):
            if partition_parts[index] != conjugate_form[index]:
                return False
        
        return True
    
    @staticmethod
    def calculate_recursive_with_memoization(target_number: int, 
                                           exact_parts: int, 
                                           memoization_table: Dict[Tuple[int, int], int]) -> int:
        """
        Recursive approach with memoization for counting self-conjugate partitions
        
        Args:
            target_number: Number to partition
            exact_parts: Exact number of parts required
            memoization_table: Dictionary for storing computed results
            
        Returns:
            Number of self-conjugate partitions of target_number with exact_parts parts
        """
        if target_number == 0 and exact_parts == 0:
            return 1
        if target_number < 0 or exact_parts <= 0:
            return 0
        if exact_parts > target_number:
            return 0
        
        memo_key = (target_number, exact_parts)
        if memo_key in memoization_table:
            return memoization_table[memo_key]
        
        valid_partitions_count = 0
        current_partition = []
        
        def generate_partitions(remaining_sum: int, remaining_parts: int, maximum_value: int) -> None:
            nonlocal valid_partitions_count
            
            if remaining_parts == 0 and remaining_sum == 0:
                temporary_partition = sorted(current_partition, reverse=True)
                
                if (len(temporary_partition) == exact_parts and 
                    SelfConjugatePartitions.validate_self_conjugate_property(temporary_partition)):
                    valid_partitions_count += 1
                return
            
            if remaining_parts <= 0 or remaining_sum <= 0:
                return
            
            for candidate_value in range(min(maximum_value, remaining_sum), 0, -1):
                current_partition.append(candidate_value)
                generate_partitions(remaining_sum - candidate_value, remaining_parts - 1, candidate_value)
                current_partition.pop()
        
        generate_partitions(target_number, exact_parts, target_number)
        memoization_table[memo_key] = valid_partitions_count
        return valid_partitions_count
    
    @staticmethod
    def calculate_dynamic_programming(target_number: int, exact_parts: int) -> int:
        """
        Dynamic programming approach for counting self-conjugate partitions
        
        Args:
            target_number: Number to partition
            exact_parts: Exact number of parts required
            
        Returns:
            Number of self-conjugate partitions of target_number with exact_parts parts
        """
        if target_number == 0 and exact_parts == 0:
            return 1
        if target_number < 0 or exact_parts <= 0:
            return 0
        if exact_parts > target_number:
            return 0
        
        dp_table = [[0] * (exact_parts + 1) for _ in range(target_number + 1)]
        dp_table[0][0] = 1
        
        for current_sum in range(1, target_number + 1):
            for current_parts in range(1, min(current_sum, exact_parts) + 1):
                valid_count = 0
                working_partition = []
                
                def generate_valid_partitions(remaining_sum: int, 
                                            remaining_parts_count: int, 
                                            maximum_allowed: int) -> None:
                    nonlocal valid_count
                    
                    if remaining_parts_count == 0 and remaining_sum == 0:
                        partition_copy = sorted(working_partition, reverse=True)
                        
                        if (len(partition_copy) == current_parts and 
                            SelfConjugatePartitions.validate_self_conjugate_property(partition_copy)):
                            valid_count += 1
                        return
                    
                    if remaining_parts_count <= 0 or remaining_sum <= 0:
                        return
                    
                    for next_value in range(min(maximum_allowed, remaining_sum), 0, -1):
                        working_partition.append(next_value)
                        generate_valid_partitions(remaining_sum - next_value, 
                                                remaining_parts_count - 1, 
                                                next_value)
                        working_partition.pop()
                
                generate_valid_partitions(current_sum, current_parts, current_sum)
                dp_table[current_sum][current_parts] = valid_count
        
        return dp_table[target_number][exact_parts]
    
    @staticmethod
    def load_parameters_from_file(file_name: str) -> Tuple[int, int]:
        """
        Read input parameters from file
        
        Args:
            file_name: Input file name
            
        Returns:
            Tuple of (target_number, exact_parts) values
        """
        try:
            with open(file_name, 'r') as input_file:
                line = input_file.readline().strip()
                numbers = line.split()
                if len(numbers) >= 2:
                    target_number = int(numbers[0])
                    exact_parts = int(numbers[1])
                    return target_number, exact_parts
                else:
                    raise ValueError("Insufficient numbers in input file")
        except FileNotFoundError:
            print(f"Error: Cannot open file {file_name}")
            exit(1)
        except ValueError as e:
            print(f"Error: Invalid input format - {e}")
            exit(1)
    
    @staticmethod
    def present_computation_results(target_number: int, 
                                  exact_parts: int, 
                                  recursive_result: int, 
                                  dp_result: int) -> None:
        """
        Display computation results with formatted output
        
        Args:
            target_number: The number being partitioned
            exact_parts: The exact number of parts required
            recursive_result: Result from recursive approach
            dp_result: Result from dynamic programming approach
        """
        print(f"Self-conjugate partitions of {target_number} with {exact_parts} parts (Recursive): {recursive_result}")
        print(f"Self-conjugate partitions of {target_number} with {exact_parts} parts (Dynamic Programming): {dp_result}")


def main() -> None:
    """Main program entry point"""
    INPUT_FILE_NAME = "input.txt"
    
    target_number, exact_parts = SelfConjugatePartitions.load_parameters_from_file(INPUT_FILE_NAME)
    
    # Compute using recursive approach with memoization
    memoization_cache = {}
    recursive_result = SelfConjugatePartitions.calculate_recursive_with_memoization(
        target_number, exact_parts, memoization_cache
    )
    
    # Compute using dynamic programming approach
    dynamic_programming_result = SelfConjugatePartitions.calculate_dynamic_programming(
        target_number, exact_parts
    )
    
    # Display results
    SelfConjugatePartitions.present_computation_results(
        target_number, exact_parts, recursive_result, dynamic_programming_result
    )


if __name__ == "__main__":
    main() 

File: Project1/Baitoan3/c/self_conjugate_partition_k_parts.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <algorithm>
#include <functional>

namespace SelfConjugatePartitions {
    
    /**
     * @brief Check if a partition is self-conjugate (equals its transpose)
     * @param partitionParts Vector of partition parts in non-increasing order
     * @return True if partition is self-conjugate, false otherwise
     */
    bool validateSelfConjugateProperty(const std::vector<int>& partitionParts) {
        int numberOfParts = static_cast<int>(partitionParts.size());
        std::vector<int> conjugateForm;
        int maximumPart = 0;
        
        for (int currentPart : partitionParts) {
            maximumPart = std::max(maximumPart, currentPart);
        }
        
        for (int rowIndex = 1; rowIndex <= maximumPart; ++rowIndex) {
            int columnCount = 0;
            for (int partValue : partitionParts) {
                if (partValue >= rowIndex) {
                    ++columnCount;
                } else {
                    break;
                }
            }
            conjugateForm.push_back(columnCount);
        }
        
        if (static_cast<int>(conjugateForm.size()) != numberOfParts) {
            return false;
        }
        
        for (int index = 0; index < numberOfParts; ++index) {
            if (partitionParts[index] != conjugateForm[index]) {
                return false;
            }
        }
        
        return true;
    }
    
    /**
     * @brief Recursive approach with memoization for counting self-conjugate partitions
     * @param targetNumber Number to partition
     * @param exactParts Exact number of parts required
     * @param memoizationTable Table for storing computed results
     * @return Number of self-conjugate partitions of targetNumber with exactParts parts
     */
    int calculateRecursiveWithMemoization(int targetNumber, 
                                         int exactParts, 
                                         std::map<std::pair<int, int>, int>& memoizationTable) {
        if (targetNumber == 0 && exactParts == 0) {
            return 1;
        }
        if (targetNumber < 0 || exactParts <= 0) {
            return 0;
        }
        if (exactParts > targetNumber) {
            return 0;
        }
        
        std::pair<int, int> memoKey = {targetNumber, exactParts};
        if (memoizationTable.count(memoKey)) {
            return memoizationTable[memoKey];
        }
        
        int validPartitionsCount = 0;
        std::vector<int> currentPartition;
        
        std::function<void(int, int, int)> generatePartitions = 
            [&](int remainingSum, int remainingParts, int maximumValue) {
                if (remainingParts == 0 && remainingSum == 0) {
                    std::vector<int> temporaryPartition = currentPartition;
                    std::sort(temporaryPartition.rbegin(), temporaryPartition.rend());
                    
                    if (static_cast<int>(temporaryPartition.size()) == exactParts && 
                        validateSelfConjugateProperty(temporaryPartition)) {
                        ++validPartitionsCount;
                    }
                    return;
                }
                
                if (remainingParts <= 0 || remainingSum <= 0) {
                    return;
                }
                
                for (int candidateValue = std::min(maximumValue, remainingSum); candidateValue >= 1; --candidateValue) {
                    currentPartition.push_back(candidateValue);
                    generatePartitions(remainingSum - candidateValue, remainingParts - 1, candidateValue);
                    currentPartition.pop_back();
                }
            };
        
        generatePartitions(targetNumber, exactParts, targetNumber);
        return memoizationTable[memoKey] = validPartitionsCount;
    }
    
    /**
     * @brief Dynamic programming approach for counting self-conjugate partitions
     * @param targetNumber Number to partition
     * @param exactParts Exact number of parts required
     * @return Number of self-conjugate partitions of targetNumber with exactParts parts
     */
    int calculateDynamicProgramming(int targetNumber, int exactParts) {
        if (targetNumber == 0 && exactParts == 0) {
            return 1;
        }
        if (targetNumber < 0 || exactParts <= 0) {
            return 0;
        }
        if (exactParts > targetNumber) {
            return 0;
        }
        
        std::vector<std::vector<int>> dpTable(targetNumber + 1, std::vector<int>(exactParts + 1, 0));
        dpTable[0][0] = 1;
        
        for (int currentSum = 1; currentSum <= targetNumber; ++currentSum) {
            for (int currentParts = 1; currentParts <= std::min(currentSum, exactParts); ++currentParts) {
                int validCount = 0;
                std::vector<int> workingPartition;
                
                std::function<void(int, int, int)> generateValidPartitions = 
                    [&](int remainingSum, int remainingPartsCount, int maximumAllowed) {
                        if (remainingPartsCount == 0 && remainingSum == 0) {
                            std::vector<int> partitionCopy = workingPartition;
                            std::sort(partitionCopy.rbegin(), partitionCopy.rend());
                            
                            if (static_cast<int>(partitionCopy.size()) == currentParts && 
                                validateSelfConjugateProperty(partitionCopy)) {
                                ++validCount;
                            }
                            return;
                        }
                        
                        if (remainingPartsCount <= 0 || remainingSum <= 0) {
                            return;
                        }
                        
                        for (int nextValue = std::min(maximumAllowed, remainingSum); nextValue >= 1; --nextValue) {
                            workingPartition.push_back(nextValue);
                            generateValidPartitions(remainingSum - nextValue, remainingPartsCount - 1, nextValue);
                            workingPartition.pop_back();
                        }
                    };
                
                generateValidPartitions(currentSum, currentParts, currentSum);
                dpTable[currentSum][currentParts] = validCount;
            }
        }
        
        return dpTable[targetNumber][exactParts];
    }
    
    /**
     * @brief Read input parameters from file
     * @param fileName Input file name
     * @return Pair of (n, k) values
     */
    std::pair<int, int> loadParametersFromFile(const std::string& fileName) {
        std::ifstream inputFile(fileName);
        if (!inputFile.is_open()) {
            std::cerr << "Error: Cannot open file " << fileName << std::endl;
            exit(1);
        }
        
        int targetNumber, exactParts;
        inputFile >> targetNumber >> exactParts;
        inputFile.close();
        
        return {targetNumber, exactParts};
    }
    
    /**
     * @brief Display computation results with formatted output
     * @param targetNumber The number being partitioned
     * @param exactParts The exact number of parts required
     * @param recursiveResult Result from recursive approach
     * @param dpResult Result from dynamic programming approach
     */
    void presentComputationResults(int targetNumber, 
                                  int exactParts, 
                                  int recursiveResult, 
                                  int dpResult) {
        std::cout << "Self-conjugate partitions of " << targetNumber 
                  << " with " << exactParts << " parts (Recursive): " << recursiveResult << std::endl;
        std::cout << "Self-conjugate partitions of " << targetNumber 
                  << " with " << exactParts << " parts (Dynamic Programming): " << dpResult << std::endl;
    }
    
} // namespace SelfConjugatePartitions

/**
 * @brief Main program entry point
 * @return Program exit status
 */
int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    
    const std::string INPUT_FILE_NAME = "input.txt";
    
    std::pair<int, int> parameters = SelfConjugatePartitions::loadParametersFromFile(INPUT_FILE_NAME);
    int targetNumber = parameters.first;
    int exactParts = parameters.second;
    
    // Clear memoization and compute using recursive approach
    std::map<std::pair<int, int>, int> memoizationCache;
    int recursiveResult = SelfConjugatePartitions::calculateRecursiveWithMemoization(
        targetNumber, exactParts, memoizationCache
    );
    
    // Compute using dynamic programming approach
    int dynamicProgrammingResult = SelfConjugatePartitions::calculateDynamicProgramming(
        targetNumber, exactParts
    );
    
    // Display results
    SelfConjugatePartitions::presentComputationResults(
        targetNumber, exactParts, recursiveResult, dynamicProgrammingResult
    );
    
    return 0;
} 